feat: add Text Fragment Composition API for i18n support

hoteye opened the pull request:
## Summary

This PR implements the Text Fragment Composition API to enable userland i18n packages to work seamlessly with Ink's reconciler, addressing the feedback from sindresorhus on enabling better i18n integration.

## Key Features

- **`composeTextFragments(fragments: TextFragment[]): string`** - Core composition function
- **`InlineTextStyle`** - Type-safe styling interface extracted from TextProps  
- **`TextFragment`** - Union type supporting both strings and styled text objects
- **Transform functions** - Support for dynamic text processing with proper indexing
- **Style compatibility** - Uses existing `colorize` and `chalk` systems with identical ordering

## Implementation Details

### Type Definitions
```typescript
export type InlineTextStyle = Pick<
    TextProps,
    | 'color' | 'backgroundColor' | 'bold' | 'italic' 
    | 'underline' | 'strikethrough' | 'inverse'
>;

export type TextFragment = string | {
    readonly text: string;
    readonly styles?: readonly InlineTextStyle[];
    readonly transform?: (text: string, index: number) => string;
};
```

### Usage Examples

**Basic Composition:**
```typescript
const result = composeTextFragments([
    "Hello ",
    { text: "World", styles: [{color: 'green', bold: true}] },
    "!"
]);
```

**i18n Integration:**
```typescript
const fragments = parseI18nTemplate(
    t('user.greeting'), // "Hello {name}, you have {count} messages"  
    {
        name: { text: userName, styles: [{color: 'green', bold: true}] },
        count: { text: messageCount.toString(), styles: [{color: 'yellow'}] }
    }
);
return <Text>{composeTextFragments(fragments)}</Text>;
```

## Technical Approach

- **Style Application Order**: Matches Text component exactly (color → backgroundColor → bold → italic → underline → strikethrough → inverse)
- **Error Handling**: Comprehensive validation with descriptive error messages
- **Performance**: Single-pass processing with efficient string concatenation
- **Compatibility**: Reuses existing Ink infrastructure (`colorize`, `chalk`)

## Test Coverage

- ✅ **28 comprehensive test cases** covering all functionality
- ✅ **Basic composition** - strings, empty arrays, mixed types
- ✅ **Style application** - all supported styles and combinations  
- ✅ **Transform functions** - with and without indexing
- ✅ **Color support** - named, hex, RGB, ANSI256 colors
- ✅ **Error handling** - invalid inputs and edge cases
- ✅ **Integration** - compatibility with existing Ink style system

## Benefits for i18n Authors

1. **Solves reconciler limitations** - Single Text component output
2. **Type safety** - Full TypeScript support with proper error messages
3. **Familiar API** - Consistent with existing Ink styling patterns
4. **Flexible input** - Supports both simple strings and complex styled objects
5. **Performance optimized** - Minimal overhead for text processing

## Addresses Issue

This implementation directly addresses [the feedback from sindresorhus](https://github.com/vadimdemedes/ink/issues/779#issuecomment-3336340187) about providing a single API for i18n packages to work with Ink's reconciler constraints.

## Test Results

All existing tests pass + 28 new tests for the new functionality:

```
✔ compose-text-fragments › composeTextFragments - basic string composition
✔ compose-text-fragments › composeTextFragments - transform and styles  
✔ compose-text-fragments › composeTextFragments - complex mixed fragments
... (25 more test cases)

28 tests passed
```

sindresorhus commented on test/compose-text-fragments.tsx:
```diff
@@ -0,0 +1,332 @@
+	t.is(result, expected);
+});
+
+test('composeTextFragments - color style', t => {
+	const fragments: TextFragment[] = [
+		{
+			text: 'Hello',
+			styles: [{color: 'red'}],
+		},
+	];
+	const result = composeTextFragments(fragments);
+	const expected = chalk.red('Hello');
+	t.is(result, expected);
+});
+
+test('composeTextFragments - background color style', t => {
+	const fragments: TextFragment[] = [
+		{
+			text: 'Hello',
+			styles: [{backgroundColor: 'blue'}],
+		},
+	];
+	const result = composeTextFragments(fragments);
+	const expected = chalk.bgBlue('Hello');
+	t.is(result, expected);
+});
```
Don't really need test for every style.

hoteye replied:
Done in [27e3fd3](https://github.com/vadimdemedes/ink/pull/780/commits/27e3fd3b57c9fb4b5f25de9b787c09b67afb9f2e)

sindresorhus commented on src/text-styles.ts:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Apply text styles to a string using the same logic as the Text component.
+ * This function centralizes the text styling logic to ensure consistency
+ * between the Text component and the composeTextFragments function.
```
Don't prefix doc comment lines with `*`.

hoteye replied:
Done in [27e3fd3](https://github.com/vadimdemedes/ink/pull/780/commits/27e3fd3b57c9fb4b5f25de9b787c09b67afb9f2e)

hoteye commented:
Hi @sindresorhus, thank you for your detailed review and feedback. After implementing all your suggestions, I've decided to convert this PR to draft for further consideration. During research, using a renderStyledText approach:

  {renderStyledText(
    t('basics.addContext'), // "Add context: Use {symbol} to specify files"
    {
      symbol: <Text bold color={theme.text.accent}>@</Text>,
      example: <Text bold color={theme.text.accent}>@src/file.ts</Text>
    },
    theme.text.primary
  )}

Translation-friendly: Complete sentences in translation files; React-native: Uses actual React components as placeholders; Simpler API: More aligned with existing i18next patterns

While our composeTextFragments implementation is technically sound and addresses all your feedback, the renderStyledText pattern appears more natural for the React/Ink ecosystem. I believe this warrants further discussion about the best approach for solving i18n fragmentation in Ink.

What are your thoughts on exploring the renderStyledText pattern instead?

hoteye commented:
Hi @sindresorhus,

I've implemented the `renderStyledText` approach in the examples directory as an alternative solution. The implementation is based on Google Gemini CLI's proven approach and includes:

**Location**: `examples/style-showcase/render-styled-text.tsx`

**Key Features**:
- Complete semantic context for translators (no fragmented strings)
- React component placeholders in translation strings  
- Full styling support compatible with Ink's Text component
- Comprehensive error handling for placeholder/styleMap mismatches
- Apache 2.0 licensed (from Google's implementation)

**Example Usage**:
```typescript
// Translation: "Shell mode: Execute commands via {symbol} (e.g., {example})"
renderStyledText(t('shellMode'), {
  symbol: <Text bold color="purple">!</Text>,
  example: <Text bold color="purple">!npm start</Text>
}, 'white')
```

**Working Demo**: `examples/style-showcase/style-showcase-i18n.tsx`

This approach solves the i18n fragmentation problem more elegantly than `composeTextFragments` by:
1. Preserving complete sentences for translators
2. Using familiar React component patterns
3. Avoiding ANSI escape code complexity

If you find this approach viable, I'd be happy to draft a new focused PR specifically for `renderStyledText`. What are your thoughts?

sindresorhus commented:
Review of `composeTextFragments`:

* **README: incorrect claim.**
  Change “Ink’s reconciler cannot handle nested Text components” to the actual constraint: `<Text>` supports *nested `<Text>` and text nodes only*; other components inside `<Text>` aren’t allowed. Link to the project note to avoid confusion.

* **README: broken examples.**
  Multiple snippets are wrong/misleading:

* **Types: align with `Text` props.**
  `InlineTextStyle.color`/`backgroundColor` are typed as `string`. Use the same unions as `Text` to preserve TS ergonomics and parity (e.g. `LiteralUnion<ForegroundColorName, string>`). You’re already importing these types in `Text.tsx`; mirror them in `text-styles.ts` and `compose-text-fragments.ts`.

* **`composeTextFragments`: inherited background footgun.**
  Right now, inherited background applies **only** if you pass at least one style object, which forces hacks like `styles: [{}]` (see test “parent background only”). Make it work without that hack: if `inheritedBackgroundColor` is provided and `styles.length === 0`, call `applyTextStyles(text, {}, inheritedBackgroundColor)` once. Update the test to drop the empty style object.

* **Docs: clarify `inheritedBackgroundColor`.**
  Explicitly state *when* to pass it (e.g., composing inside a component that must honor a parent `<Text>` background while keeping the wrapper unstyled) and that it’s not auto-detected. Point to the combined background/inverse test as the behavioral reference.

sindresorhus commented:
> During research, I found that a renderStyledText approach might be more intuitive:

From what I can tell, `renderStyledText` does not require any changes in Ink, so it should definitely not be in Ink. It can just be a separate package.

sindresorhus commented:
If you don't need this PR, I'm not sure it's worth adding at all? It creates extra maintenance, and there is no clear use-case.

I'm also not super confident in the code. It's clearly vibe coded.

sindresorhus commented on readme.md:
```diff
@@ -1460,6 +1461,211 @@ Type: `number`
+// Simple text composition
+const fragments = [
+  'Status: ',
+  { text: 'Online', styles: [{ color: 'green', bold: true }] }
```
Format it properly over multiple lines. Same with other examples.

hoteye commented:
Thank you for your review. I have decided to close this PR as it has been repeatedly reverted to a draft during the development process. I believe that some aspects of its functionality are insufficient or unnecessary at this time. If needed, we can revisit this in the future with a more refined approach.
